pkg/store/consul.go

```diff
@@ -97,9 +97,14 @@ func (s *Consul) setEphemeralTTL(ttl time.Duration) error {
 
 // CreateEphemeralSession creates the a global session
 // once that is used to delete keys at node failure
-func (s *Consul) createEphemeralSession() error {
+func (s *Consul) createEphemeralSession(key string) error {
 	s.Lock()
 	defer s.Unlock()
+
+	// Delete the key first (erase any attached session for that key)
+	// FIXME recover gracefully with the existing global session
+	s.Delete(key)
```

Comment:
It deletes it a single time although it does not work as expected: each subsequent key going through the Put will stay with the old session (it thus only work truly with a single key, which is bad). A better solution is to get back the existing session but at the cost of the Manager not seeing the "change" of the Agent leaving and coming back to life.

I will keep this PR open and give it a shot.